### PR TITLE
Feature/106 sep point

### LINKED ISSUE #106: sepPoint fsw module
**Describe your use case**
This module expands the capabilities of oneAxisSolarArrayPoint by adding the option to prioritize a Sun-constrained body-frame axis, which must be kept within a certain angular distance from the Sun. For readability, this new module has specific input modules and parameters, which make it such that the module can only be used in SEP-Pointing mode.
This module prioritizes: 1) thrust axis alignment, 2) solar array incidence, 3) Sun-constrained direction. Priorities 2) and 3) can be swapped.

### PR DESCRIPTION
**Pull request by rcalaon**
* **Tickets addressed:** bsk-106
* **Review:** By commit  
* **Merge strategy:** Merge (no squash)  

## Description
This branch delivers a module sepPoint that is a refactored version of the already existing oneAxisSolarArrayPoint. This new version focuses on a single flight mode and univocally defines the input messages and input parameters to the module. The new module incorporates a new feature that enables the user to choose which pointing requirements to prioritize, after the main thruster pointing requirement is met. In the first commit, a library of utility function is created inside the attGuidance folder, which contains functions used in this module and in similar future modules. Commit 2 initializes the module files. Commit 3 populates the Reset function to check that all the inputs are connected correctly. Commit 4 contains the bulk of the module, where the Update function is populated, and the output message is computed and written. Commits 5 and 6 contain the unit tests and the documentation, respectively.

## Verification
Two unit tests are provided, to test the correct functionality of the module with the two different sets of priorities.

## Documentation
Documentation is provided to describe how to use the module.

## Future work
Future work will involve the development of more similar modules that use similar math, but satisfy the pointing requirements for different flight modes.

### CODE REVIEW

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+ */
+
+#include "attitudePointingLibrary.h"
+#include "architecture/utilities/linearAlgebra.h"
+#include "architecture/utilities/rigidBodyKinematics.h"
+#include "architecture/utilities/astroConstants.h"
+#include <math.h>
+
+/*! This constructor initializes an Input structure for BSpline interpolation */
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: :large_orange_diamond: Looks like a copy and paste comment.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    }
+}
+
+/*! Generic destructor */
+SolutionSpace::~SolutionSpace() = default;
+
+/*! Define the contains method */
+int SolutionSpace::numberOfZeros() {return this->zeros;}
+
+/*! Define the contains method */
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: :large_orange_diamond: More copy and paste comments.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+#include "architecture/utilities/astroConstants.h"
+#include <math.h>
+
+/*! This constructor initializes an Input structure for BSpline interpolation */
+SolutionSpace::SolutionSpace()
+{
+    this->isEmpty = true;
+}
+
+SolutionSpace::SolutionSpace(double A, double B, double C, double tol)
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp:30]: :large_orange_diamond: We should add some docstring (doxygen) comments to theses methods which describe what they do and what their parameters are.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+                this->sup =  M_PI;
+            }
+            else {
+                this->isEmpty = true;
+            }
+            this->zeros = 0;
+        }
+    }
+    // compute max and min of the function
+    double psi1, psi2, y1, y2;
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: :large_orange_diamond: It's good practice to define one variable per line in these situations.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+
+/*! This constructor initializes an Input structure for BSpline interpolation */
+SolutionSpace::SolutionSpace()
+{
+    this->isEmpty = true;
+}
+
+SolutionSpace::SolutionSpace(double A, double B, double C, double tol)
+{
+    this->tol = tol;
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: :large_orange_diamond: We assign `this->tol` here but we don't end up using it anywhere in the class. Do we have other plans for it?

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+ */
+
+#include "attitudePointingLibrary.h"
+#include "architecture/utilities/linearAlgebra.h"
+#include "architecture/utilities/rigidBodyKinematics.h"
+#include "architecture/utilities/astroConstants.h"
+#include <math.h>
+
+/*! This constructor initializes an Input structure for BSpline interpolation */
+SolutionSpace::SolutionSpace()
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: :small_red_triangle_down: I think we might be able to get rid of this constructor. Is there a use case where one would want an uninitialized problem space? If so, it doesn't look like we have the getters/setters to carry out initialization after construction.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+ WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+ */
+
+#include "attitudePointingLibrary.h"
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp:20]: :large_orange_diamond: Let's chat about what to call this set class.

File: src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp [hunk 8]
```diff
@@ -65,6 +65,151 @@ void SepPoint::Reset(uint64_t CurrentSimNanos)
...
+    double RN[3][3];
+    m33MultM33(RB, BN, RN);
+
+    /*! compute reference MRP */
+    double sigma_RN[3];
+    C2MRP(RN, sigma_RN);
+    v3Copy(sigma_RN, attRefOut.sigma_RN);
+
+    /*! compute reference angular rates and accelerations via finite differences */
+    double omega_RN_R[3], omegaDot_RN_R[3];
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp]: :large_orange_diamond: Better practice to declare a single variable per line/statement.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.h [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+//! @brief The SolutionSpace class contains the solutions of the nonlinear inequality
+//! |Ax^2 + Bx + C| / (1+x^2) >= 0
+class SolutionSpace {
+public:
+    SolutionSpace();
+    SolutionSpace(double A, double B, double C, double tol);
+    ~SolutionSpace();
+
+    bool   isEmpty;                  //!< determines whether the solution space is empty
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.h]: :large_orange_diamond: Should `isEmpty` should be made private and accessed through an `isEmpty()`? It seems like this is a variable that should be set internally only based upon the computation of the solution space. 

If it needs to be set we could also add a `setEmpty(bool val)`

File: src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp [hunk 10]
```diff
@@ -65,6 +65,151 @@ void SepPoint::Reset(uint64_t CurrentSimNanos)
...
+
+        case sunConstrAxisAlign :
+            double k = cos(this->beta);
+            SolutionSpace sunConstAxisSolutions(D-k, E, F-k, epsilon);
+            if (sunConstAxisSolutions.isEmpty) {
+                double psi = sunConstAxisSolutions.returnAbsMin(1);
+                v3Scale(psi, e_psi, PRV_psi);
+            }
+            else {
+                if (solarArraySolutions.numberOfZeros() == 2) {
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp]: :large_orange_diamond: It seems like we could reduce the number of indentations here and change this to an `else if` and result in the same logic.

File: src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp [hunk 11]
```diff
@@ -65,6 +65,151 @@ void SepPoint::Reset(uint64_t CurrentSimNanos)
...
+    m33MultM33(RB, BN, RN);
+
+    /*! compute reference MRP */
+    double sigma_RN[3];
+    C2MRP(RN, sigma_RN);
+    v3Copy(sigma_RN, attRefOut.sigma_RN);
+
+    /*! compute reference angular rates and accelerations via finite differences */
+    double omega_RN_R[3], omegaDot_RN_R[3];
+    finiteDifferencesRatesAndAcc(sigma_RN, this->sigma_RN_1, this->sigma_RN_2, CurrentSimNanos,
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp]: :large_orange_diamond: I think we should drop each parameter onto a new line. like the following minus the extra funky indent that is there from my copy and paste.

```
finiteDifferencesRatesAndAcc(sigma_RN,
                                 this->sigma_RN_1,
                                 this->sigma_RN_2,
                                 CurrentSimNanos,
                                 this->T1NanoSeconds,
                                 this->T2NanoSeconds,
                                 this->callCount,
                                 omega_RN_R,
                                 omegaDot_RN_R);
```

File: src/fswAlgorithms/attGuidance/sepPoint/_UnitTest/test_sepPoint0.py [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+    d = np.cross(a1Hat_B, [6, -5, 4])
+    d = d / np.linalg.norm(d)
+    
+    DCM1 = rbk.PRV2C(a * alpha)
+    DCM2 = rbk.PRV2C(d * delta)
+
+    hHat_N = np.matmul(DCM1, rHat_SB_N)             # required thrust direction in inertial frame, at an angle alpha from rHat_SB_N
+    hHat_B = np.matmul(DCM2, a1Hat_B)               # required thrust direction in inertial frame, at an angle alpha from rHat_SB_N
+
+    a2Hat_B = [0, 1, 0]
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/sepPoint/_UnitTest/test_sepPoint0.py]: :small_red_triangle_down: You can remove these unused variables `a2Hat_B`, `testFailCount`, `testMessages`

File: src/fswAlgorithms/attGuidance/sepPoint/_UnitTest/test_sepPoint0.py [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+# the code is less accurate around this point;
+ang = np.linspace(0, np.pi, 10, endpoint=False)
+ang = list(ang)
+
+
+@pytest.mark.parametrize("alpha", ang)
+@pytest.mark.parametrize("delta", ang)
+@pytest.mark.parametrize("alignmentPriority", [0])
+@pytest.mark.parametrize("accuracy", [1e-12])
+def test_sepPointTestFunction(show_plots, alpha, delta, alignmentPriority, accuracy):
```
Comment (patkenneally) [inline src/fswAlgorithms/attGuidance/sepPoint/_UnitTest/test_sepPoint0.py]: You could move these decorators down to the below function and rename it `test_` . This extra function isn't needed now that we've gotten rid of the testMessage/testFailures stuff

**Review by patkenneally — CHANGES_REQUESTED** [review]

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp (hunk 1 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: Fixed it.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp (hunk 2 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: Fixed it.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp (hunk 3 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp:30]: Added some comments.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp (hunk 5 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: You're right, I removed this class property altogether.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp (hunk 6 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: Right. I don't see the need for that at the moment. I removed it.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp (hunk 4 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.cpp]: Fixed it.

File: src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.h (hunk 9 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/_GeneralModuleFiles/attitudePointingLibrary.h]: I implemented this change

File: src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp (hunk 8 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp]: Done

File: src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp (hunk 11 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp]: Done

File: src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp (hunk 10 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/sepPoint/sepPoint.cpp]: Fixed it.

File: src/fswAlgorithms/attGuidance/sepPoint/_UnitTest/test_sepPoint0.py (hunk 12 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/sepPoint/_UnitTest/test_sepPoint0.py]: Done

File: src/fswAlgorithms/attGuidance/sepPoint/_UnitTest/test_sepPoint0.py (hunk 13 above)
Comment (rcalaon) [reply src/fswAlgorithms/attGuidance/sepPoint/_UnitTest/test_sepPoint0.py]: Done

**Review by patkenneally — APPROVED** [review]

**Pull request merged**